router/widen-lemma.gobra

```diff
@@ -0,0 +1,1532 @@
+// Copyright 2022 ETH Zurich
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +gobra
+
+package router
+
+import (
+	sl "verification/utils/slices"
+	"verification/io"
+	. "verification/utils/definitions"
+	"verification/dependencies/encoding/binary"
+	"github.com/scionproto/scion/pkg/slayers/path"
+	"github.com/scionproto/scion/pkg/slayers/path/scion"
+)
+
+ghost 
+requires 0 <= length && length <= len(raw)
+requires length > 4
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R50)
+requires acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R50)
+requires unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), R50) in
+	let hdr := binary.BigEndian.Uint32(raw[0:4]) in
+	let metaHdr := scion.DecodedFrom(hdr) in
+	let seg1 := int(metaHdr.SegLen[0]) in
+	let seg2 := int(metaHdr.SegLen[1]) in
+	let seg3 := int(metaHdr.SegLen[2]) in
+	let base := scion.Base{metaHdr,
+		numInfoFields(seg1, seg2, seg3),
+		seg1 + seg2 + seg3} in
+	metaHdr.InBounds() &&
+	0 < metaHdr.SegLen[0] &&
+	base.ValidCurrInfSpec() &&
+	base.ValidCurrHfSpec() &&
+	len(raw) >= pktLen(seg1, seg2, seg3)
+requires unfolding acc(sl.AbsSlice_Bytes(raw[:length], 0, len(raw[:length])), R50) in
+	let hdr := binary.BigEndian.Uint32(raw[:length][0:4]) in
+	let metaHdr := scion.DecodedFrom(hdr) in
+	let seg1 := int(metaHdr.SegLen[0]) in
+	let seg2 := int(metaHdr.SegLen[1]) in
+	let seg3 := int(metaHdr.SegLen[2]) in
+	let base := scion.Base{metaHdr,
+		numInfoFields(seg1, seg2, seg3),
+		seg1 + seg2 + seg3} in
+	metaHdr.InBounds() &&
+	0 < metaHdr.SegLen[0] &&
+	base.ValidCurrInfSpec() &&
+	base.ValidCurrHfSpec() &&
+	len(raw[:length]) >= pktLen(seg1, seg2, seg3)
```

Comment:
are these preconditions really necessary? I am afraid that this will make using the lemma impractical. Additionally, don't we have a pure function for this? these expressions are very large and repeating them along the proof makes everything much harder to read